node/pkg/driver/node.go

```diff
@@ -613,9 +627,95 @@ func (d *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 }
 
 func (d *NodeService) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	goid_info.SetAdditionalIDInfo(req.VolumeId)
+	volumeId := req.VolumeId
+	volumePath := req.VolumePath
+	volumePathWithHostPrefix := d.NodeUtils.GetPodPath(req.VolumePath)
+	isFSVolume := true
+	goid_info.SetAdditionalIDInfo(volumeId)
 	defer goid_info.DeleteAdditionalIDInfo()
-	return nil, status.Error(codes.Unimplemented, "NodeGetVolumeStats is not implemented yet")
+
+	if volumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume ID must be provided")
+	}
+	if volumePath == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume Path must be provided")
+	}
+
+	isPathExists := d.NodeUtils.IsPathExists(volumePathWithHostPrefix)
+	if !isPathExists {
+		return nil, status.Errorf(codes.NotFound, "volume path %q does not exists", volumePath)
+	}
+	isBlock, err := d.isBlock(volumePathWithHostPrefix)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to determine if %q is block device: %s", volumePath, err)
+	}
+	if isBlock {
+		isFSVolume = false
+	}
+	isMounted, err := d.IsMounted(volumePathWithHostPrefix, isFSVolume)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to check if volume path %q is mounted", volumePath)
+	}
+	if !isMounted {
+		return nil, status.Errorf(codes.NotFound, "volume path %q is not mounted", volumePath)
+	}
+	volumeStats, err := d.getVolumeStats(volumePathWithHostPrefix)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to get capacity statistics for volume path %q: %s", volumePath, err)
+	}
+	return &csi.NodeGetVolumeStatsResponse{
+		Usage: []*csi.VolumeUsage{
```

Comment:
don't we want volume_condition too?
I can see it's Must if we support "VOLUME_CONDITION node capability" but any reason why not do it anyway?

Reply:
`A Node plugin MUST implement this RPC call if it has GET_VOLUME_STATS node capability or VOLUME_CONDITION node capability`
this is what they write in the spec, so it's enough to add GET_VOLUME_STATS node capability, no?

Reply:
we don't support [volume health monitoring](https://kubernetes.io/blog/2021/04/16/volume-health-monitoring-alpha-update/#node-side-volume-health-monitoring), I couldn't find other CSI drivers that support this capacility, I don't think it's trivial and it's still alpha anyway

Reply:
ok, tnx.